### ISSUE TITLE
INVALID_RESOURCE_ID but with valid payment

### ISSUE THREAD

**Issue by truefusion** [paypal/PayPal-PHP-SDK#937]
> for Internal Tracking: PPPLMER-39145
### General information

* SDK/Library version: 1.11.0
* Environment: Production
* `PayPal-Debug-ID` values: 584e6b4ca79f8
* Language, language version, and OS: PHP 7, Debian Linux

### Issue description

This error occurred when a user attempted to make a purchase using the "paypal" payment method. PayPal returned an INVALID_RESOURCE_ID error, but our PayPal transaction records show that the payment did indeed go through. (I recognize this may not be the issue of the API itself but some internal problem.)

**Comment by xiaoleih41** [comment]
I saw two payment attempts made within seconds. There might have been some internal issue for the first payment failure  but the second transaction went through.

**Comment by brandonkruse** [comment]
Had the same issues happen with us today. We process about 3,000 PayPal payments/day and we had this happen to about 50 of them.

When we go to execute the payment, we get a 404. Note that this DOES withdraw money from the account of the payer

1. The customer was redirect to PayPal to accept the payment
2. She accepted and was redirected back to our application
3. We called the API /execute method for the payment and in the logs we got:

[25-09-2017 11:19:45] PayPal\Core\PayPalHttpConnection: INFO    : GET https://api.paypal.com/v1/payments/payment/PAY-********339LHESX5I
[25-09-2017 11:19:46] PayPal\Core\PayPalHttpConnection: INFO    : Response Status   : 200
[25-09-2017 11:19:46] PayPal\Core\PayPalHttpConnection: INFO    : POST https://api.paypal.com/v1/payments/payment/PAY-********339LHESX5I/execute
[25-09-2017 11:19:49] PayPal\Core\PayPalHttpConnection: INFO    : Response Status   : 404
[25-09-2017 11:19:49] PayPal\Core\PayPalHttpConnection: ERROR   : Got Http response code 404 when accessing https://api.paypal.com/v1/payments/payment/PAY-********339LHESX5I/execute. {"name":"INVALID_RESOURCE_ID","message":"The requested resource ID was not found","information_link":"https://developer.paypal.com/docs/api/payments/#errors","debug_id":"881aecfb5f5f"}

4. This exception caused the order not to be marked as paid in our system, however the customer had the amount withdrawn from their PayPal account

**Comment by brandonkruse** [comment]
I opened ticket #170925-000412

**Comment by xelhark** [comment]
I had this problem too. 

Just to shed some light on the problem, I noticed that when this happens, there's also a few secondary effects:

1) Even after the transaction is purchased, the payment returned from the API will always remain in a "created" state, and will contain no sale or transaction data (making it impossible to tie the payment to its sale / transaction)

2) The Sale object generated by paypal will also be orphan. Meaning that the field "parent_payment_id" will not even appear on the json of the sale object associated with the transaction id (this again, makes it pretty hard to find a way to tie the transaction object with the payment).

3) I noticed that it looks like when this happens, sometimes, a second execute call (which will happen a lot since users gets an error message) will even result in duplicated transactions. The only way to stop this is to make sure the setting "Block accidental payments" is set to block payments with duplicated invoice IDs (my selling tools > block payments).

As a workaround, I managed to store the payment data in my db (along with other data) and inject my internal payment object id in the "custom" data of the transaction when creating the payment.

This will allow me to always be able to retrieve the data associated with the payment...


..Except this also seem to be happening during recurring payments as well, and there's no custom data there..

What I'm doing in this case is storing my payment id in the first characters of the description, but this is a very ugly (and hackish) solution.. The users will also be able to see that id and it's not the best user experience, since that field has even a 128 character limit..
